### ISSUE TITLE
3.7.2 build fails

### ISSUE THREAD

**Issue by barjac** [dl1ksv/gr-fcdproplus#14]
Installing /home/baz/rpmbuild/SRPMS/gr-fcdproplus-3.7.2-1.mga6.src.rpm
Executing(%prep): /bin/sh -e /home/baz/rpmbuild/tmp/rpm-tmp.KWzy1Z
+ umask 022
+ cd /home/baz/rpmbuild/BUILD
+ '[' 1 -eq 1 ']'
+ '[' 1 -eq 1 ']'
+ '[' 1 -eq 1 ']'
+ cd /home/baz/rpmbuild/BUILD
+ rm -rf gr-fcdproplus-3.7.2
+ /usr/bin/gzip -dc /home/baz/rpmbuild/SOURCES/gr-fcdproplus-3.7.2.tar.gz
+ /usr/bin/tar -xof -
+ STATUS=0
+ '[' 0 -ne 0 ']'
+ cd gr-fcdproplus-3.7.2
+ /usr/bin/chmod -Rf a+rX,u+w,g-w,o-w .
+ rm -rf lib/hid
+ dos2unix lib/fcdcmd.h
dos2unix: converting file lib/fcdcmd.h to Unix format...
+ exit 0
Executing(%build): /bin/sh -e /home/baz/rpmbuild/tmp/rpm-tmp.71BPUa
+ umask 022
+ cd /home/baz/rpmbuild/BUILD
+ cd gr-fcdproplus-3.7.2
+ '[' 1 -eq 1 ']'
+ '[' 1 -eq 1 ']'
+ CFLAGS='-O2 -g -pipe -Wformat -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -fstack-protector --param=ssp-buffer-size=4'
+ export CFLAGS
+ CXXFLAGS='-O2 -g -pipe -Wformat -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -fstack-protector --param=ssp-buffer-size=4'
+ export CXXFLAGS
+ FFLAGS='-O2 -g -pipe -Wformat -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -fstack-protector --param=ssp-buffer-size=4'
+ export FFLAGS
+ LDFLAGS=' -Wl,--as-needed -Wl,--no-undefined -Wl,-z,relro -Wl,-O1 -Wl,--build-id -Wl,--enable-new-dtags'
+ export LDFLAGS
+ mkdir -p build
+ cd build
+ /usr/bin/cmake .. -DCMAKE_INSTALL_PREFIX:PATH=/usr -DCMAKE_INSTALL_LIBDIR:PATH=/usr/lib64 -DCMAKE_INSTALL_LIBEXECDIR:PATH=/usr/libexec -DCMAKE_INSTALL_SYSCONFDIR:PATH=/etc -DINCLUDE_INSTALL_DIR:PATH=/usr/include -DLIB_INSTALL_DIR:PATH=/usr/lib64 -DSYSCONF_INSTALL_DIR:PATH=/etc -DSHARE_INSTALL_PREFIX:PATH=/usr/share -DCMAKE_BUILD_TYPE=RelWithDebInfo -DCMAKE_CXX_FLAGS_RELWITHDEBINFO:STRING=-DNDEBUG -DCMAKE_C_FLAGS_RELWITHDEBINFO:STRING=-DNDEBUG -DLIB_SUFFIX=64 -DCMAKE_SKIP_RPATH:BOOL=ON -DCMAKE_VERBOSE_MAKEFILE:BOOL=ON '-DCMAKE_MODULE_LINKER_FLAGS=-Wl,--as-needed  -Wl,-z,relro -Wl,-O1 -Wl,--build-id -Wl,--enable-new-dtags' -DBUILD_SHARED_LIBS:BOOL=ON -DBUILD_STATIC_LIBS:BOOL=OFF -DENABLE_DOXYGEN=on -DGR_PKG_DOC_DIR=/usr/share/doc/gr-fcdproplus ..
-- The CXX compiler identification is GNU 5.3.1
-- The C compiler identification is GNU 5.3.1
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Boost version: 1.60.0
-- Found the following Boost libraries:
--   filesystem
--   system
-- Found PkgConfig: /usr/bin/pkg-config (found version "0.29") 
Checking for GNU Radio Module: RUNTIME
-- Checking for module 'gnuradio-runtime'
--   Found gnuradio-runtime, version 3.7.9.1
 * INCLUDES=/usr/include
 * LIBS=/lib64/libgnuradio-runtime.so;/lib64/libgnuradio-pmt.so
-- Found GNURADIO_RUNTIME: /lib64/libgnuradio-runtime.so;/lib64/libgnuradio-pmt.so  
GNURADIO_RUNTIME_FOUND = TRUE
Checking for GNU Radio Module: AUDIO
-- Checking for module 'gnuradio-audio'
--   Found gnuradio-audio, version 3.7.9.1
 * INCLUDES=/usr/include
 * LIBS=/lib64/libgnuradio-audio.so;/lib64/libgnuradio-runtime.so;/lib64/libgnuradio-pmt.so
-- Found GNURADIO_AUDIO: /lib64/libgnuradio-audio.so;/lib64/libgnuradio-runtime.so;/lib64/libgnuradio-pmt.so  
GNURADIO_AUDIO_FOUND = TRUE
Checking for GNU Radio Module: BLOCKS
-- Checking for module 'gnuradio-blocks'
--   Found gnuradio-blocks, version 3.7.9.1
 * INCLUDES=/usr/include
 * LIBS=/lib64/libgnuradio-blocks.so;/lib64/libgnuradio-runtime.so;/lib64/libgnuradio-pmt.so
-- Found GNURADIO_BLOCKS: /lib64/libgnuradio-blocks.so;/lib64/libgnuradio-runtime.so;/lib64/libgnuradio-pmt.so  
GNURADIO_BLOCKS_FOUND = TRUE
Checking for GNU Radio Module: PMT
-- Checking for module 'gnuradio-runtime'
--   Found gnuradio-runtime, version 3.7.9.1
 * INCLUDES=/usr/include
 * LIBS=/lib64/libgnuradio-runtime.so;/lib64/libgnuradio-pmt.so
-- Found GNURADIO_PMT: /lib64/libgnuradio-runtime.so;/lib64/libgnuradio-pmt.so  
GNURADIO_PMT_FOUND = TRUE
CMake Error at CMakeLists.txt:94 (include):
  include could not find load file:

    GrVersion

CMake Error at CMakeLists.txt:98 (include):
  include could not find load file:

    GrPlatform

-- Found ALSA: /usr/lib64/libasound.so (found version "1.1.0") 
CMake Warning at CMakeLists.txt:115 (find_package):
  By not providing "FindUSB.cmake" in CMAKE_MODULE_PATH this project has
  asked CMake to find a package configuration file provided by "USB", but
  CMake did not find one.

  Could not find a package configuration file provided by "USB" with any of
  the following names:

    USBConfig.cmake
    usb-config.cmake

  Add the installation prefix of "USB" to CMAKE_PREFIX_PATH or set "USB_DIR"
  to a directory containing one of the above files.  If "USB" provides a
  separate development package or SDK, be sure it has been installed.

CMake Error at CMakeLists.txt:117 (message):
  libusb required to compile fcdproplus

-- Configuring incomplete, errors occurred!
See also "/home/baz/rpmbuild/BUILD/gr-fcdproplus-3.7.2/build/CMakeFiles/CMakeOutput.log".
error: Bad exit status from /home/baz/rpmbuild/tmp/rpm-tmp.71BPUa (%build)

#############################################
We have the same BuildRequires as for all previous versions:-
BuildRequires:    cmake
BuildRequires:    pkgconfig(libusb)
BuildRequires:    pkgconfig(gnuradio-runtime)
BuildRequires:    pkgconfig(gnuradio-audio)
BuildRequires:    pkgconfig(gnuradio-blocks)
BuildRequires:    gnuradio-pmt-devel
BuildRequires:    hidapi-devel
BuildRequires:    gnuradio-utils
BuildRequires:    boost-devel
BuildRequires:    dos2unix
BuildRequires:    doxygen
BuildRequires:    graphviz
BuildRequires:    swig
BuildRequires:    alsa-lib-devel

Our gnuradio-utils package provdes:
 /usr/lib64/cmake/gnuradio/GrPlatform.cmake
/usr/lib64/cmake/gnuradio/FindUSB.cmake
/usr/lib64/cmake/gnuradio/GrVersion.cmake

**Comment by dl1ksv** [comment]
Thanks for testing. Did you try to build against an existing gnuradio installation ? Normaly the gnuradio cmake modules are located in ${CMAKE_INSTALL_PREFIX}/lib/cmake/gnuradio. That's the location I'm searching the cmake modules.
In your case the location seems to be /usr/lib64/cmake/gnuradio. Is this always true ? Is there a way to get this location, or have I to provide cmake Module to find the location of these modules ?

**Comment by barjac** [comment]
We build in a chroot with a minimal base system plus the BuildRequires for the package being built.
CMake files for gnuradio are packaged in gnuradio-utils which is a BuildRequire, so they will always be in: %{_libdir}/cmake/gnuradio,  which is /usr/lib64/cmake/gnuradio  for a x86_64 build and  /usr/lib/cmake/gnuradio for a i586 build.

**Comment by dl1ksv** [comment]
I just pushed a module that should find the Modules directory an fix this problem. Could you please check if it works for you?
For late use: Does the gnuradio-utils package provide all gnuradio cmake Modules ?

**Comment by barjac** [comment]
Thanks , I will test later by making a patch for 3.7.2
Yes all the modules are there.
Now what happened here?: https://github.com/dl1ksv/gr-fcdproplus/releases
We seem to be back at 0.1.0 ?

**Comment by dl1ksv** [comment]
I removed the v3.7.2 tag. As soon as I know that  the latest commit fixes your problem I'll tag this commit as v3.7.2 as this version should work on all distros and I would this change to my other gnuradio modules.

**Comment by barjac** [comment]
OK I tested it with a patch made for 3.7.2 using
 git diff 6c8e09674e4ab56b9c > patch
The cmake modules are now found OK, but there is now another issue.
We delete the lib/hid directory to force use of the system hidapi package before starting the build and this has always worked in the past.
Now it appears to fail looking for the source, after it has already found the system lib.
-- ================ Build Summary =========================
-- Building gr-fcdproplus : 3.7.2 / 3.7.2 for Linux 
-- Using CMAKE Module path: /home/baz/rpmbuild/BUILD/gr-fcdproplus-3.7.2/cmake/Modules;/usr/lib64/cmake/gnuradio
-- GNURADIO INCLUDES IN   : /usr/include
-- Using install prefix   : /usr
-- GNURADIO LIBS          : /lib64/libgnuradio-audio.so;/lib64/libgnuradio-runtime.so;/lib64/libgnuradio-pmt.so
-- System Hidapi Lib /usr/lib64/libhidapi-libusb.so is used
-- ========================================================
-- Configuring done
CMake Error at lib/CMakeLists.txt:61 (add_library):
  Cannot find source file:

    /home/baz/rpmbuild/BUILD/gr-fcdproplus-3.7.2/lib/hid/hid.c

  Tried extensions .c .C .c++ .cc .cpp .cxx .m .M .mm .h .hh .h++ .hm .hpp
  .hxx .in .txx

CMake Error: CMake can not determine linker language for target: gnuradio-fcdproplus
CMake Error: Cannot determine link language for target "gnuradio-fcdproplus".
-- Generating done

**Comment by dl1ksv** [comment]
This is a bug. I'll fix this.

**Comment by dl1ksv** [comment]
Should be fixed with  latest commit https://github.com/dl1ksv/gr-fcdproplus/commit/269592450007543dee72efe56e8384ef4539a822

**Comment by barjac** [comment]
Yes that issue is fixed.
We are now getting a mix of lib versions - see the full build log.
http://mtf.no-ip.co.uk/pub/linux/barjac/distrib/cauldron/x86_64/log/gr-fcdproplus-3.7.2-1.mga6.src.rpm/build.0.20160311175927.log
Note that I temporarily removed the libs from the package files list so we could see what the build was producing:
   /usr/lib64/libgnuradio-fcdproplus.so
   /usr/lib64/libgnuradio-fcdproplus.so.0.0.0
   /usr/lib64/libgnuradio-fcdproplus.so.3.7.2

I guess this is related to the versioning change.

**Comment by dl1ksv** [comment]
Yes, the multiple libs are a result of the versioning change. Last week I tried to get some recommendations how OOT-modules should be versioned. But I failed. At the moment it's a 'missing' feature. So 'ill do a feature request.
But I found the function that gnuradio uses to name it's libs (GR_LIBRARY_FOO).
The result is
libgnuradio-\<component\>-\<libver\>.so.0.0.0
and three different links to this lib
libgnurradio-\<component\>.so
libgnuradio-\<component\>-\<libver\>.so
libgnuradio-\<component\>-\<libver\>.so.0

So I hope it's a good decision to use this function for versioning OOT-modules, too.
I do this with the [latest commit](https://github.com/dl1ksv/gr-fcdproplus/commit/7cef4fa4586e6e630d6dbd495aa8fdb55b9aa9d4)

**Comment by barjac** [comment]
Yes that now builds OK for me, and cmake modules are found
-- Installing: /home/baz/rpmbuild/BUILDROOT/gr-fcdproplus-3.7.2-0.rc1.4.mga6.x86_64/usr/lib64/libgnuradio-fcdproplus-3.7.2.so.0.0.0
-- Installing: /home/baz/rpmbuild/BUILDROOT/gr-fcdproplus-3.7.2-0.rc1.4.mga6.x86_64/usr/lib64/libgnuradio-fcdproplus-3.7.2.so
-- Installing: /home/baz/rpmbuild/BUILDROOT/gr-fcdproplus-3.7.2-0.rc1.4.mga6.x86_64/usr/lib64/libgnuradio-fcdproplus.so
-- Installing: /home/baz/rpmbuild/BUILDROOT/gr-fcdproplus-3.7.2-0.rc1.4.mga6.x86_64/usr/lib64/libgnuradio-fcdproplus-3.7.2.so.0

**Comment by barjac** [comment]
Ah - I had not noticed that there is an extra symlink being created:
libgnuradio-fcdproplus.so
when we now also have:
libgnuradio-fcdproplus-3.7.2.so
Not sure if this is acceptable, but I deleted the first one from the package with no apparent ill-effects.
gr-osmosdr and gqrx build against it without errors.
I don't have fcd hardware to test.

**Comment by dl1ksv** [comment]
Well, that's how gnuradio names it libs. I don't know why. But if you want to delete some links I'd prefer the libgnuradio-fcdproplus-3.7.2.so as pkg-config references   libgnuradio-fdcproplus.so

**Comment by barjac** [comment]
OK no problem - it's fine with both so for me this is fixed thanks for all your help.
Call here by the way is G4MKT and the package is for Mageia.
73
Barry

**Comment by dl1ksv** [comment]
Good to hear. Then I'll close this issue. I'll update my other gnuradio modules. If no further modifications are necessary I'll tag this commit as release.
Perhaps we'll meet on the bands.
73 -- Volker
